src/neural/blas/network_blas.cc

```diff
@@ -310,16 +311,134 @@ void BlasComputation<use_eigen>::ComputeBlocking() {
           SELU,  // SELU activation for attention head.
           head_buffer.data());
 
-      for (auto layer : weights_.pol_encoder) {
-        // TODO: support encoder heads.
-        throw Exception(
-            "Eigen/Blas backend doesn't support encoder heads yet.");
-      }
       const size_t policy_d_model = weights_.ip2_pol_b.size();
-      std::vector<float> head_buffer2(largest_batch_size * policy_d_model *
+      const size_t max_channel_size = weights_.pol_encoder.size() > 0
+                            ? weights_.pol_encoder[0].ffn.dense1_b.size() // DFF size
+                            : policy_d_model;
+      std::vector<float> head_buffer2(largest_batch_size * max_channel_size *
+                                      kSquares);
+      std::vector<float> head_buffer3(largest_batch_size * max_channel_size *
                                       kSquares);
-      std::vector<float> head_buffer3(largest_batch_size * policy_d_model *
+      std::vector<float> head_buffer4(largest_batch_size * max_channel_size *
                                       kSquares);
+      std::vector<float> temp_buffer1(policy_d_model * kSquares);
+      std::vector<float> temp_buffer2(policy_d_model * kSquares);
+      std::vector<float> temp_buffer3(policy_d_model * kSquares);
+
+      for (auto layer : weights_.pol_encoder) {
+        // Q
+        FullyConnectedLayer<use_eigen>::Forward1D(
+            batch_size * kSquares, embedding_size, layer.mha.q_b.size(),
+            head_buffer.data(), layer.mha.q_w.data(),
+            layer.mha.q_b.data(), NONE, head_buffer2.data());
+        // K
+        FullyConnectedLayer<use_eigen>::Forward1D(
+            batch_size * kSquares, embedding_size, layer.mha.k_b.size(),
+            head_buffer.data(), layer.mha.k_w.data(),
+            layer.mha.k_b.data(), NONE, head_buffer3.data());
+        // V
+        FullyConnectedLayer<use_eigen>::Forward1D(
+            batch_size * kSquares, embedding_size, layer.mha.v_b.size(),
+            head_buffer.data(), layer.mha.v_w.data(),
+            layer.mha.v_b.data(), NONE, head_buffer4.data());
+
+        // MHA (Q, K, V)
+        const int d_model = layer.mha.q_b.size();
+        const int heads = weights_.pol_encoder_head_count;
+        const int depth = d_model / heads;
+        const float scaling = 1.0f / sqrtf(d_model);
+
+        // MHA is done per batch since there's a fourth dimension introduced.
+        for (auto batch = size_t{0}; batch < batch_size; batch++) {
+          auto batchStart = batch * kSquares * d_model;
+
+          // Reshape and transpose for each head.
+          const float * Q = temp_buffer1.data();
+          const float * K = temp_buffer2.data();
+          const float * V = temp_buffer3.data();
+          
+          for (auto head = size_t{0}; head < heads; head++) {
```

Comment:
Make it `int` instead of `size_t` to avoid warnings (and in the next 2 loops using `heads`).